File: src/arithmetic_mixed_numbers_hard.py
"""
arithmetic - mixed_numbers (hard)
Generated: 2026-02-22T03:42:19.259372
"""

from problem_utils import *

def generate():
    # For harder mixed number problems (1600-1900 ELO), we'll create problems involving:
    # - Complex operations with multiple mixed numbers
    # - Multi-step word problems
    # - Operations requiring multiple conversions
    # - Problems with larger numbers or multiple fractions
    
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Complex multiplication/division chain with mixed numbers
        whole1, num1, den1 = randint(2, 5), randint(1, 7), randint(2, 8)
        while num1 >= den1:
            num1, den1 = randint(1, 7), randint(2, 8)
        
        whole2, num2, den2 = randint(1, 4), randint(1, 5), randint(2, 6)
        while num2 >= den2:
            num2, den2 = randint(1, 5), randint(2, 6)
        
        whole3, num3, den3 = randint(1, 3), randint(1, 4), randint(2, 5)
        while num3 >= den3:
            num3, den3 = randint(1, 4), randint(2, 5)
        
        frac1 = Rational(whole1 * den1 + num1, den1)
        frac2 = Rational(whole2 * den2 + num2, den2)
        frac3 = Rational(whole3 * den3 + num3, den3)
        
        # Either multiply three or multiply two and divide by third
        if randint(0, 1) == 0:
            result = frac1 * frac2 * frac3
            operation = "multiplication"
            question = f"${whole1}\\frac{{{num1}}}{{{den1}}} \\times {whole2}\\frac{{{num2}}}{{{den2}}} \\times {whole3}\\frac{{{num3}}}{{{den3}}}$"
            step1 = f"Convert to improper fractions: $\\frac{{{frac1.p}}}{{{frac1.q}}} \\times \\frac{{{frac2.p}}}{{{frac2.q}}} \\times \\frac{{{frac3.p}}}{{{frac3.q}}}$"
            step2 = f"Multiply numerators and denominators: $\\frac{{{frac1.p * frac2.p * frac3.p}}}{{{frac1.q * frac2.q * frac3.q}}}$"
        else:
            result = (frac1 * frac2) / frac3
            operation = "mixed"
            question = f"${whole1}\\frac{{{num1}}}{{{den1}}} \\times {whole2}\\frac{{{num2}}}{{{den2}}} \\div {whole3}\\frac{{{num3}}}{{{den3}}}$"
            step1 = f"Convert to improper fractions: $\\frac{{{frac1.p}}}{{{frac1.q}}} \\times \\frac{{{frac2.p}}}{{{frac2.q}}} \\div \\frac{{{frac3.p}}}{{{frac3.q}}}$"
            step2 = f"Multiply by reciprocal: $\\frac{{{frac1.p}}}{{{frac1.q}}} \\times \\frac{{{frac2.p}}}{{{frac2.q}}} \\times \\frac{{{frac3.q}}}{{{frac3.p}}}$"
        
        if result.q == 1:
            answer_str = str(result.p)
            step3 = f"Simplify to get ${result.p}$"
        else:
            whole_part = result.p // result.q
            num_part = result.p % result.q
            if whole_part > 0:
                answer_str = f"{whole_part}\\frac{{{num_part}}}{{{result.q}}}"
                step3 = f"Simplify and convert to mixed number: ${answer_str}$"
            else:
                answer_str = f"\\frac{{{result.p}}}{{{result.q}}}"
                step3 = f"Simplify: ${answer_str}$"
        
        return problem(
            question=f"Calculate: {question}",
            answer=answer_str,
            difficulty=(1600, 1750),
            topic="arithmetic/mixed_numbers",
            solution=steps(step1, step2, step3)
        )
    
    elif problem_type == 2:
        # Word problem with multiple mixed number operations
        whole1, num1, den1 = randint(3, 8), randint(1, 3), randint(4, 4)
        whole2, num2, den2 = randint(2, 5), randint(1, 2), randint(3, 3)
        whole3, num3, den3 = randint(1, 3), randint(1, 1), randint(2, 2)
        
        frac1 = Rational(whole1 * den1 + num1, den1)
        frac2 = Rational(whole2 * den2 + num2, den2)
        frac3 = Rational(whole3 * den3 + num3, den3)
        
        result = frac1 - frac2 + frac3
        
        question = f"A carpenter has a board that is ${whole1}\\frac{{{num1}}}{{{den1}}}$ feet long. He cuts off ${whole2}\\frac{{{num2}}}{{{den2}}}$ feet, then adds a piece that is ${whole3}\\frac{{{num3}}}{{{den3}}}$ feet long. What is the total length of the board now?"
        
        step1 = f"Convert to improper fractions: $\\frac{{{frac1.p}}}{{{frac1.q}}} - \\frac{{{frac2.p}}}{{{frac2.q}}} + \\frac{{{frac3.p}}}{{{frac3.q}}}$"
        
        # Find common denominator
        from math import gcd
        lcm_denom = (frac1.q * frac2.q * frac3.q) // gcd(gcd(frac1.q, frac2.q), frac3.q)
        common_denom = frac1.q * frac2.q // gcd(frac1.q, frac2.q)
        common_denom = common_denom * frac3.q // gcd(common_denom, frac3.q)
        
        step2 = f"Find common denominator and compute: $\\frac{{{result.p}}}{{{result.q}}}$"
        
        if result.q == 1:
            answer_str = str(result.p)
            step3 = f"The board is ${result.p}$ feet long"
        else:
            whole_part = result.p // result.q
            num_part = result.p % result.q
            if whole_part > 0:
                answer_str = f"{whole_part}\\frac{{{num_part}}}{{{result.q}}}"
                step3 = f"Convert to mixed number: ${answer_str}$ feet"
            else:
                answer_str = f"\\frac{{{result.p}}}{{{result.q}}}"
                step3 = f"The board is ${answer_str}$ feet long"
        
        return problem(
            question=question,
            answer=answer_str,
            difficulty=(1650, 1800),
            topic="arithmetic/mixed_numbers",
            solution=steps(step1, step2, step3)
        )
    
    elif problem_type == 3:
        # Complex addition/subtraction with three mixed numbers requiring common denominator
        whole1, num1, den1 = randint(4, 9), randint(1, 4), randint(5, 5)
        whole2, num2, den2 = randint(3, 7), randint(2, 5), randint(6, 6)
        while num2 >= den2:
            num2 = randint(2, 5)
        whole3, num3, den3 = randint(2, 6), randint(1, 7), randint(8, 8)
        while num3 >= den3:
            num3 = randint(1, 7)
        
        frac1 = Rational(whole1 * den1 + num1, den1)
        frac2 = Rational(whole2 * den2 + num2, den2)
        frac3 = Rational(whole3 * den3 + num3, den3)
        
        # Randomly choose operations
        if randint(0, 1) == 0:
            result = frac1 + frac2 - frac3
            question = f"${whole1}\\frac{{{num1}}}{{{den1}}} + {whole2}\\frac{{{num2}}}{{{den2}}} - {whole3}\\frac{{{num3}}}{{{den3}}}$"
        else:
            result = frac1 - frac2 + frac3
            question = f"${whole1}\\frac{{{num1}}}{{{den1}}} - {whole2}\\frac{{{num2}}}{{{den2}}} + {whole3}\\frac{{{num3}}}{{{den3}}}$"
        
        step1 = f"Convert to improper fractions: $\\frac{{{frac1.p}}}{{{frac1.q}}}, \\frac{{{frac2.p}}}{{{frac2.q}}}, \\frac{{{frac3.p}}}{{{frac3.q}}}$"
        
        # Common denominator
        from math import gcd
        common_denom = frac1.q * frac2.q // gcd(frac1.q, frac2.q)
        common_denom = common_denom * frac3.q // gcd(common_denom, frac3.q)
        
        step2 = f"Find common denominator ({common_denom}) and compute"
        
        if result.q == 1:
            answer_str = str(result.p)
            step3 = f"Result: ${result.p}$"
        else:
            whole_part = result.p // result.q
            num_part = result.p % result.q
            if whole_part > 0:
                answer_str = f"{whole_part}\\frac{{{num_part}}}{{{result.q}}}"
                step3 = f"Simplify to mixed number: ${answer_str}$"
            else:
                answer_str = f"\\frac{{{result.p}}}{{{result.q}}}"
                step3 = f"Result: ${answer_str}$"
        
        return problem(
            question=f"Calculate: {question}",
            answer=answer_str,
            difficulty=(1600, 1750),
            topic="arithmetic/mixed_numbers",
            solution=steps(step1, step2, step3)
        )
    
    else:
        # Division problem with large mixed numbers
        whole1, num1, den1 = randint(5, 12), randint(1, 5), randint(6, 6)
        while num1 >= den1:
            num1 = randint(1, 5)
        whole2, num2, den2 = randint(2, 4), randint(1, 3), randint(4, 4)
        while num2 >= den2:
            num2 = randint(1, 3)
        
        frac1 = Rational(whole1 * den1 + num1, den1)
        frac2 = Rational(whole2 * den2 + num2, den2)
        
        result = frac1 / frac2
        
        question = f"${whole1}\\frac{{{num1}}}{{{den1}}} \\div {whole2}\\frac{{{num2}}}{{{den2}}}$"
        
        step1 = f"Convert to improper fractions: $\\frac{{{frac1.p}}}{{{frac1.q}}} \\div \\frac{{{frac2.p}}}{{{frac2.q}}}$"
        step2 = f"Multiply by reciprocal: $\\frac{{{frac1.p}}}{{{frac1.q}}} \\times \\frac{{{frac2.q}}}{{{frac2.p}}} = \\frac{{{frac1.p * frac2.q}}}{{{frac1.q * frac2.p}}}$"
        
        if result.q == 1:
            answer_str = str(result.p)
            step3 = f"Simplify: ${result.p}$"
        else:
            whole_part = result.p // result.q
            num_part = result.p % result.q
            if whole_part > 0:
                answer_str = f"{whole_part}\\frac{{{num_part}}}{{{result.q}}}"
                step3 = f"Simplify to mixed number: ${answer_str}$"
            else:
                answer_str = f"\\frac{{{result.p}}}{{{result.q}}}"
                step3 = f"Simplify: ${answer_str}$"
        
        return problem(
            question=f"Calculate: {question}",
            answer=answer_str,
            difficulty=(1700, 1850),
            topic="arithmetic/mixed_numbers",
            solution=steps(step1, step2, step3)
        )

emit(generate())

File: src/arithmetic_mixed_numbers_hard.jl
# arithmetic - mixed_numbers (hard)
# Generated: 2026-03-08T19:48:56.981663

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("arithmetic/mixed_numbers")
    
    problem_type = rand(1:4)
    
    if problem_type == 1
        # Complex multiplication with large mixed numbers
        whole1 = randint(8, 25)
        num1 = randint(1, 19)
        den1 = randint(num1 + 1, 24)
        
        whole2 = randint(7, 20)
        num2 = randint(1, 17)
        den2 = randint(num2 + 1, 22)
        
        # Convert to improper fractions
        imp1 = whole1 + num1 // den1
        imp2 = whole2 + num2 // den2
        
        product = imp1 * imp2
        
        ans_whole = floor(Int, product)
        ans_num = numerator(product - ans_whole)
        ans_den = denominator(product - ans_whole)
        
        if ans_num == 0
            answer = string(ans_whole)
        else
            answer = "$(ans_whole) $(ans_num)/$(ans_den)"
        end
        
        problem(
            question="Calculate: \\($(whole1)\\frac{$(num1)}{$(den1)} \\times $(whole2)\\frac{$(num2)}{$(den2)}\\). Express your answer as a mixed number in lowest terms.",
            answer=answer,
            difficulty=(1850, 2100),
            solution=steps(
                "Convert to improper fractions: \\(\\frac{$(whole1 * den1 + num1)}{$(den1)} \\times \\frac{$(whole2 * den2 + num2)}{$(den2)}\\)",
                "Multiply numerators and denominators: \\(\\frac{$((whole1 * den1 + num1) * (whole2 * den2 + num2))}{$(den1 * den2)}\\)",
                "Simplify and convert to mixed number",
                "Answer: \\($(ans_whole)\\frac{$(ans_num)}{$(ans_den)}\\)"
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Complex division with strategic cancellation
        whole1 = randint(10, 30)
        num1 = randint(2, 15)
        den1 = randint(num1 + 2, 20)
        
        whole2 = randint(3, 12)
        num2 = randint(1, 11)
        den2 = randint(num2 + 1, 16)
        
        imp1 = whole1 + num1 // den1
        imp2 = whole2 + num2 // den2
        
        quotient = imp1 / imp2
        
        ans_whole = floor(Int, quotient)
        ans_num = numerator(quotient - ans_whole)
        ans_den = denominator(quotient - ans_whole)
        
        if ans_num == 0
            answer = string(ans_whole)
        else
            answer = "$(ans_whole) $(ans_num)/$(ans_den)"
        end
        
        problem(
            question="Calculate: \\($(whole1)\\frac{$(num1)}{$(den1)} \\div $(whole2)\\frac{$(num2)}{$(den2)}\\). Express your answer as a mixed number in lowest terms.",
            answer=answer,
            difficulty=(1900, 2200),
            solution=steps(
                "Convert to improper fractions: \\(\\frac{$(whole1 * den1 + num1)}{$(den1)} \\div \\frac{$(whole2 * den2 + num2)}{$(den2)}\\)",
                "Multiply by reciprocal: \\(\\frac{$(whole1 * den1 + num1)}{$(den1)} \\times \\frac{$(den2)}{$(whole2 * den2 + num2)}\\)",
                "Simplify and convert to mixed number",
                "Answer: \\($(ans_whole)\\frac{$(ans_num)}{$(ans_den)}\\)"
            ),
            time=200
        )
        
    elseif problem_type == 3
        # Chain operations requiring multiple conversions
        whole1 = randint(5, 18)
        num1 = randint(1, 13)
        den1 = randint(num1 + 1, 16)
        
        whole2 = randint(4, 15)
        num2 = randint(1, 11)
        den2 = randint(num2 + 1, 14)
        
        whole3 = randint(3, 12)
        num3 = randint(1, 9)
        den3 = randint(num3 + 1, 12)
        
        imp1 = whole1 + num1 // den1
        imp2 = whole2 + num2 // den2
        imp3 = whole3 + num3 // den3
        
        result = (imp1 * imp2) - imp3
        
        ans_whole = floor(Int, result)
        ans_num = numerator(result - ans_whole)
        ans_den = denominator(result - ans_whole)
        
        if ans_num == 0
            answer = string(ans_whole)
        else
            answer = "$(ans_whole) $(ans_num)/$(ans_den)"
        end
        
        problem(
            question="Evaluate: \\($(whole1)\\frac{$(num1)}{$(den1)} \\times $(whole2)\\frac{$(num2)}{$(den2)} - $(whole3)\\frac{$(num3)}{$(den3)}\\). Express your answer as a mixed number in lowest terms.",
            answer=answer,
            difficulty=(2000, 2300),
            solution=steps(
                "Convert all to improper fractions",
                "First multiply: \\(\\frac{$(whole1 * den1 + num1)}{$(den1)} \\times \\frac{$(whole2 * den2 + num2)}{$(den2)}\\)",
                "Then subtract \\(\\frac{$(whole3 * den3 + num3)}{$(den3)}\\) using common denominators",
                "Simplify and convert to mixed number",
                "Answer: \\($(ans_whole)\\frac{$(ans_num)}{$(ans_den)}\\)"
            ),
            time=240
        )
        
    else
        # Complex word problem with mixed numbers
        rate_whole = randint(12, 28)
        rate_num = randint(1, 7)
        rate_den = choice([2, 4, 8])
        
        time_whole = randint(3, 9)
        time_num = randint(1, 5)
        time_den = choice([2, 4, 6])
        
        rate = rate_whole + rate_num // rate_den
        time_val = time_whole + time_num // time_den
        
        distance = rate * time_val
        
        ans_whole = floor(Int, distance)
        ans_num = numerator(distance - ans_whole)
        ans_den = denominator(distance - ans_whole)
        
        if ans_num == 0
            answer = string(ans_whole)
        else
            answer = "$(ans_whole) $(ans_num)/$(ans_den)"
        end
        
        problem(
            question="A cyclist travels at a constant speed of \\($(rate_whole)\\frac{$(rate_num)}{$(rate_den)}\\) miles per hour for \\($(time_whole)\\frac{$(time_num)}{$(time_den)}\\) hours. How many miles does the cyclist travel? Express your answer as a mixed number in lowest terms.",
            answer=answer,
            difficulty=(1850, 2150),
            solution=steps(
                "Use distance = rate × time",
                "Convert to improper fractions: \\(\\frac{$(rate_whole * rate_den + rate_num)}{$(rate_den)} \\times \\frac{$(time_whole * time_den + time_num)}{$(time_den)}\\)",
                "Multiply: \\(\\frac{$((rate_whole * rate_den + rate_num) * (time_whole * time_den + time_num))}{$(rate_den * time_den)}\\)",
                "Simplify and convert to mixed number",
                "Answer: \\($(ans_whole)\\frac{$(ans_num)}{$(ans_den)}\\) miles"
            ),
            time=210
        )
    end
end